src/librustc_save_analysis/external_data.rs

```diff
@@ -64,6 +64,102 @@ impl SpanData {
     }
 }
 
+/// Represent an arbitrary attribute on a code element
+#[derive(Clone, Debug, RustcEncodable)]
+pub struct Attribute {
+    value: AttributeItem,
+    span: SpanData,
+}
+
+impl Lower for ast::Attribute {
+    type Target = Attribute;
+
+    fn lower(self, tcx: TyCtxt) -> Attribute {
+        Attribute {
+            value: self.value.lower(tcx),
+            span: SpanData::from_span(self.span, tcx.sess.codemap()),
+        }
+    }
+}
+
+impl Lower for Vec<ast::Attribute> {
+    type Target = Vec<Attribute>;
+
+    fn lower(self, tcx: TyCtxt) -> Vec<Attribute> {
+        self.into_iter().map(|x| x.lower(tcx)).collect()
+    }
+}
+
+/// A single item as part of an attribute
+#[derive(Clone, Debug, RustcEncodable)]
+pub struct AttributeItem {
+    name: LitKind,
+    kind: AttributeItemKind,
+    span: SpanData,
+}
+
+impl Lower for ast::MetaItem {
+    type Target = AttributeItem;
+
+    fn lower(self, tcx: TyCtxt) -> AttributeItem {
+        AttributeItem {
+            name: LitKind::Str(self.name, StrStyle::Cooked),
+            kind: self.node.lower(tcx),
+            span: SpanData::from_span(self.span, tcx.sess.codemap()),
+        }
+    }
+}
+
+impl Lower for ast::NestedMetaItem {
+    type Target = AttributeItem;
+
+    fn lower(self, tcx: TyCtxt) -> AttributeItem {
+        match self.node {
+            ast::NestedMetaItemKind::MetaItem(item) => item.lower(tcx),
+            ast::NestedMetaItemKind::Literal(lit) => {
+                AttributeItem {
+                    name: lit.node,
+                    kind: AttributeItemKind::Literal,
+                    span: SpanData::from_span(lit.span, tcx.sess.codemap()),
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, RustcEncodable)]
+pub enum AttributeItemKind {
+    /// Word meta item.
+    ///
+    /// E.g. `test` as in `#[test]`
+    Literal,
+    /// Name value meta item.
+    ///
+    /// E.g. `feature = "foo"` as in `#[feature = "foo"]`
+    NameValue(LitKind, SpanData),
+    /// List meta item.
+    ///
+    /// E.g. the `derive(..)` as in `#[derive(..)]`
+    List(Vec<AttributeItem>),
```

Comment:
I feel like this is too much structure for save-analysis - will we need it all in the RLS? Could we do something simpler?